cmd/thanos/receive.go

```diff
@@ -440,6 +440,28 @@ func runReceive(
 		})
 	}
 
+	level.Debug(logger).Log("msg", "setting up periodic top metrics collection")
+	{
+		topMetricNumSeries := promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "thanos_receive_top_metric_num_series",
+			Help: "Number of series in top metric.",
+		}, []string{"tenant", "metric_name"})
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(func() error {
+			return runutil.Repeat(conf.topMetricsUpdateInterval, ctx.Done(), func() error {
+				level.Error(logger).Log("msg", "getting top metrics")
+				for _, ts := range dbs.TenantStats(conf.numTopMetricsPerTenant, labels.MetricName) {
+					for _, ms := range ts.Stats.IndexPostingStats.CardinalityMetricsStats {
```

Comment:
is the return value not a reference so no read lock are needed? maybe use `go -race` to check